libgit/autogit_manager.go

```diff
@@ -0,0 +1,397 @@
+// `dstDir/srcRepo` in the TLF `dstTLF`. `dstDir` must already exist
+// in `dstTLF`.
+//
+// It returns a channel that, when closed, indicates the clone request
+// has finished (though not necessarily successfully).  The caller may
+// have to sync from the server to ensure they are see the changes,
+// however.
+//
+// If the cloned directory doesn't exist yet, this function creates a
+// placeholder "CLONING" file to let users know the clone is happening
+// in the background.
+//
+// Note that if there's already data in `dstDir/srcRepo`, this
+// tramples it, so the caller must ensure that they are requesting a
+// clone from the same repo/branch as the existing repo that might
+// have already been there.
+func (am *AutogitManager) Clone(
+	ctx context.Context, srcTLF *libkbfs.TlfHandle, srcRepo, branchName string,
+	dstTLF *libkbfs.TlfHandle, dstDir string) (
+	doneCh <-chan struct{}, err error) {
+	am.log.CDebugf(ctx, "Autogit clone request from %s/%s:%s to %s/%s",
+		srcTLF.GetCanonicalPath(), srcRepo, branchName,
+		dstTLF.GetCanonicalPath(), dstDir)
+	defer func() {
+		am.deferLog.CDebugf(ctx, "Clone request processed: %+v", err)
+	}()
+
+	dstFS, err := libfs.NewFS(
+		ctx, am.config, dstTLF, dstDir, "", keybase1.MDPriorityNormal)
+	if err != nil {
+		return nil, err
+	}
+
+	// Take dst lock and create "CLONING" file if needed.
+	lockFile, err := dstFS.Create(autogitLockName(srcRepo))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		closeErr := lockFile.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+	err = lockFile.Lock()
+	if err != nil {
+		return nil, err
+	}
+
+	err = dstFS.MkdirAll(srcRepo, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	dstRepoFS, err := dstFS.Chroot(srcRepo)
+	if err != nil {
+		return nil, err
+	}
+
+	fis, err := dstRepoFS.ReadDir("")
+	if err != nil {
+		return nil, err
+	}
+	if len(fis) == 0 {
+		err = am.makeCloningFile(ctx, dstRepoFS, srcTLF, srcRepo, branchName)
+		if err != nil {
+			return nil, err
+		}
+		// Sync the CLONING file before starting the reset.
+		err = lockFile.Unlock()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	req := resetReq{
+		srcTLF, srcRepo, branchName, dstTLF, dstDir, make(chan struct{}),
+	}
+	return am.queueReset(ctx, req)
+}
+
+// Pull queues a request to pull the `branchName` branch of the
+// `srcRepo` repo from the TLF `srcTLF`, into a subdirectory named
+// `dstDir/srcRepo` in the TLF `dstTLF`. `dstDir/srcRepo` must already
+// exist in `dstTLF`.
+//
+// It returns a channel that, when closed, indicates the pull request
+// has finished (though not necessarily successfully).  The caller may
+// have to sync from the server to ensure they are see the changes,
+// however.
+//
+// Note that this tramples any data that was previously in
+// `dstDir/srcRepo`, so the caller must ensure that they are
+// requesting a pull from the correct repo/branch as the existing repo
+// that was already there.
```

Comment:
I'm not planning to expose branches at first to the user, but I'm thinking the UI will have non-master branch names as part of the pathname somewhere.  So you'd never let a user switch a branch on a given checkout, it would just do a whole new checkout.

That said, if somehow a caller passed in a different branch with the same `dstDir`, it should work fine -- go-git's `Reset` should just take care of it.  The branch name just gets resolved to a commit hash, and that's what the reset uses.